src/emulator/shadergen/src/shadergen_common.cpp

```diff
@@ -0,0 +1,114 @@
+#include <shadergen/functions.h>
+
+#include <gxm/functions.h>
+#include <gxm/types.h>
+#include <shadergen/profile.h>
+#include <shadergen/types.h>
+#include <util/log.h>
+
+#include <fstream>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+namespace gxp {
+
+const char *log_parameter_semantic(const SceGxmProgramParameter &parameter) {
+    SceGxmParameterSemantic semantic = static_cast<SceGxmParameterSemantic>(parameter.semantic);
+
+    switch (semantic) {
+    // clang-format off
+    case SCE_GXM_PARAMETER_SEMANTIC_NONE: return "NONE";
+    case SCE_GXM_PARAMETER_SEMANTIC_ATTR: return "ATTR";
+    case SCE_GXM_PARAMETER_SEMANTIC_BCOL: return "BCOL";
+    case SCE_GXM_PARAMETER_SEMANTIC_BINORMAL: return "BINORMAL";
+    case SCE_GXM_PARAMETER_SEMANTIC_BLENDINDICES: return "BLENDINDICES";
+    case SCE_GXM_PARAMETER_SEMANTIC_BLENDWEIGHT: return "BLENDWEIGHT";
+    case SCE_GXM_PARAMETER_SEMANTIC_COLOR: return "COLOR";
+    case SCE_GXM_PARAMETER_SEMANTIC_DIFFUSE: return "DIFFUSE";
+    case SCE_GXM_PARAMETER_SEMANTIC_FOGCOORD: return "FOGCOORD";
+    case SCE_GXM_PARAMETER_SEMANTIC_NORMAL: return "NORMAL";
+    case SCE_GXM_PARAMETER_SEMANTIC_POINTSIZE: return "POINTSIZE";
+    case SCE_GXM_PARAMETER_SEMANTIC_POSITION: return "POSITION";
+    case SCE_GXM_PARAMETER_SEMANTIC_SPECULAR: return "SPECULAR";
+    case SCE_GXM_PARAMETER_SEMANTIC_TANGENT: return "TANGENT";
+    case SCE_GXM_PARAMETER_SEMANTIC_TEXCOORD: return "TEXCOORD";
+    default: return "UNKNOWN";
+        // clang-format on
+    }
+}
+
+void log_parameter(const SceGxmProgramParameter &parameter) {
+    std::string category;
+    switch (parameter.category) {
+    case SCE_GXM_PARAMETER_CATEGORY_ATTRIBUTE:
+        category = "Vertex attribute";
+        break;
+    case SCE_GXM_PARAMETER_CATEGORY_UNIFORM:
+        category = "Uniform";
+        break;
+    case SCE_GXM_PARAMETER_CATEGORY_SAMPLER:
+        category = "Sampler";
+        break;
+    case SCE_GXM_PARAMETER_CATEGORY_AUXILIARY_SURFACE:
+        category = "Auxiliary surface";
+        break;
+    case SCE_GXM_PARAMETER_CATEGORY_UNIFORM_BUFFER:
+        category = "Uniform buffer";
+        break;
+    default:
+        category = "Unknown type";
+        break;
+    }
+    LOG_DEBUG("{}: name:{:s} semantic:{} type:{:d} component_count:{} container_index:{}",
+        category, parameter_name_raw(parameter), log_parameter_semantic(parameter), parameter.type, uint8_t(parameter.component_count), log_hex(uint8_t(parameter.container_index)));
+}
+
+const SceGxmProgramParameter *program_parameters(const SceGxmProgram &program) {
```

Comment:
I'm pretty sure some of these functions are in gxm library.

Reply:
Good catch, didn't see them. Gotta get used to there not being a singular `gxm.cpp` anymore.